File: Problem_2 K_diff_pairs.py
# // Time Complexity : O(nlogn)
# // Space Complexity :O(n)
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : No
#
#
# // Your code here along with comments explaining your approach
# Using 2 pointers approach

class Solution:
    def findPairs(self, nums: List[int], k: int) -> int:
        nums = sorted(nums)         #Sort the numbers
        left = 0
        right = 1
        result = 0
        while left < len(nums) and right < len(nums):
            if left == right or nums[right] - nums[left] < k:
                right += 1
            elif nums[right] - nums[left] > k:
                left += 1
            else:
                left += 1
                result += 1
                #duplicate pairs are not counted repeatedly
                while left < len(nums) and nums[left] == nums[left - 1]:
                    #Increment left pointer as long as left points to number which is equal to the previous one.
                    left += 1
        return result

File: Problem2_KDiffPairs.cs
// Time Complexity : O(n^2)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes

public class Solution {
    public int FindPairs(int[] nums, int k) {
        int count = 0;
        Array.Sort(nums); //Sort the array
        for(int i = 0; i < nums.Length; i++) {
            if(i != 0 && nums[i] == nums[i-1]) { // No need to process same number more than once 
                continue;
            }
            for(int j = i + 1; j < nums.Length; j++) {
                int diff = Math.Abs(nums[i] - nums[j]); // find the difference between ith and jth numbers
                if(diff == k) {
                    //If the difference is 'k', increment the count and move on to next ith number
                    count++;
                    break;
                }
            }
        }
        return count;
    }
}